analyzers/src/SonarAnalyzer.Common/Rules/Utilities/SymbolReferenceAnalyzerBase.cs

```diff
@@ -148,7 +151,10 @@ private SymRefInfo GetSymRefInfo(SyntaxToken token, SemanticModel semanticModel)
             return null;
         }
 
-        public class SymRefInfo
+        private static bool HasTooManyTokens(SyntaxTree syntaxTree) =>
+            syntaxTree.GetRoot().DescendantTokens().Count() > TokenCountThreshold;
```

Comment:
`CreateMessage` above already does `DescendantTokens`. We can avoid 2nd evaluation if we put the condition there and don't run the foreach...
```
var tokens = syntaxTree.GetRoot().DescendantTokens();
if(tokens.Length <= TokenCountThreshold)
{
  foreach...

}
return symbolReferenceInfo;
```

Reply:
good catch! from this POV, we should do perf testing with this and before this before merging on `master`. To see the effect on "normal" projects.

Reply:
FYI: I've already did some tests to check the performance for `DescendantTokens` and it seems to be relatively cheap. For a tree with 334009 tokens it takes ~00:00:00.0591375

Reply:
Thanks.

Reply:
I prefer not to update this method. If we do the check here then we will create empty `SymbolReferenceInfo` which we will not need (or return null which is not great).

Doing the check in `ShouldGenerateMetrics` is better in my opinion since it will prevent the `CreateMessage` calls if they are not needed.